rand_core/src/block.rs

```diff
@@ -364,14 +367,34 @@ where
         self.half_used = !self.half_used;
         self.index += self.half_used as usize;
 
-        // Index as if this is a u32 slice.
+        let index = if cfg!(target_endian = "little") {
+            index
+        } else {
+            index ^ 1
+        };
+
+        let results = self.results.as_ref();
+        // Convert the `&[u64]` into an equivalent `&[u32]`, which has twice as
+        // many elements. The conversion is sound because `u64` never has a
+        // smaller alignment requirement than `u32`.
+        let ptr = results.as_ptr() as *const u32;
+        // We are guaranteed that this multiplication can't overflow. If it did
+        // overflow, that would mean that there were at least 2^(word size)
+        // 32-bit values in the slice, and thus 4 * 2^(word size) bytes. There
+        // can only ever be at most 2^(word size) - 1 addressable bytes in
+        // memory at a time, so we could never be in this situation.
+        let len = results.len() * 2;
+
         unsafe {
-            let results = &*(self.results.as_ref() as *const [u64] as *const [u32]);
-            if cfg!(target_endian = "little") {
-                *results.get_unchecked(index)
-            } else {
-                *results.get_unchecked(index ^ 1)
-            }
+            let results = core::slice::from_raw_parts(ptr, len);
+            // Note that this bounds check is required for soundness. There's no
+            // guarantee that `<<R as BlockRngCore>::Results as AsRef>::as_ref`,
+            // which is supplied by code outside of this crate, returns a slice
+            // of any particular length, or even that two subsequent calls to it
+            // return slices of the same length. Thus, it would be unsound to
+            // assume that this index is in-bounds and to elide the bounds
+            // check.
+            results[index]
```

Comment:
Could we make all of the code safe at this point?

Reply:
Yes, we can by getting `u64` from the cached block, shifting it to 32 bits if one half was already used, and casting it to `u32`.

Reply:
I think this is the best approach. Maybe let's just convert `index` to an index into the `u64` slice and shift by `32 * self.half_used` bits.

Reply:
I opened  #1160 for this.